include/openenclave/advanced/mallinfo.h

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+/**
+ * @file mallinfo.h
+ *
+ * This file defines an interface that pluggable allocators can implement.
+ * See
+ * https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/Mallinfo.md
+ *
+ */
+
+#ifndef OE_ADVANCED_MALLINFO_H
+#define OE_ADVANCED_MALLINFO_H
+
+#include <openenclave/bits/types.h>
+
+/**
+ * @cond IGNORE
+ */
+OE_EXTERNC_BEGIN
+
+/**
+ * @endcond
+ */
+
+typedef struct _oe_mallinfo
+{
+    /// Maximum number of bytes that can be allocated in total
+    size_t max_total_heap_size;
+    /// Number of bytes allocated at the moment.
+    size_t current_allocated_heap_size;
+    /// Highest value reached by `current_allocated_heap_size` during execution.
+    size_t peak_allocated_heap_size;
+} oe_mallinfo_t;
+
+/**
+ * Obtain current memory usage statistics.
+ *
+ * Users may make this call on allocators that support it, to find out how much
+ * memory can be allocated in total, how much is allocated at the moment, and
+ * the high watermark of allocation so far.
+ *
+ * @param[out] info An oe_mallinfo_t struct, to be populated by the allocator.
+ *
+ * @returns 0 on success.
```

Comment:
What does it return if the allocator doesn't support it?

Reply:
The design spec specifies non-zero, I can update this description. My intention was to leave it deliberately unspecified, and allow allocators to potentially define their own error codes if necessary.

In practice though, I believe the only realistic error value would be OE_NOT_IMPLEMENTED or something equivalent.

I've also wondered if it should be an oe_result_t, but that seemed overloaded enough already. If someone has a stronger opinion about this, I am happy to make the necessary adjustments.

Reply:
Yes I think it should be oe_result_t rather than a new error code space.

Reply:
Ok, I've made the change on both PRs.